# Movies Tonight Part 3: SQL DML

In [1]:
%load_ext sql
%sql sqlite:///Movies.db

'Connected: @Movies.db'

1. **Load the `DATASET.csv` file into the `Movies.db` database.**  
    - Import the CSV file to create a new table called `DATASET`. Follow the example [here](http://www.sqlitetutorial.net/sqlite-import-csv/).
    - Use a [`PRAGMA table-info` pseudo-query](https://www.sqlite.org/pragma.html#pragma_table_info) to inspect the design of the new table. [How are we going to handle dates](https://www.sqlite.org/datatype3.html)?

In [3]:
%%sql
SELECT * FROM DATASET LIMIT 10;

 * sqlite:///Movies.db
Done.


[('Akarakian Theatres Moreno 4 Cinemas', 'The Intersection Of Alessandro + Perris Blvds, Moreno Valley', '(909) 485-2899', 'Associate, The', '4:20 PM', 'PG-13', 'A', 'Austin Pendleton'),
 ('Akarakian Theatres Moreno 4 Cinemas', 'The Intersection Of Alessandro + Perris Blvds, Moreno Valley', '(909) 485-2899', 'Associate, The', '4:20 PM', 'PG-13', 'A', 'Bebe Neuwirth'),
 ('Akarakian Theatres Moreno 4 Cinemas', 'The Intersection Of Alessandro + Perris Blvds, Moreno Valley', '(909) 485-2899', 'Associate, The', '4:20 PM', 'PG-13', 'A', 'Dianne Wiest'),
 ('Akarakian Theatres Moreno 4 Cinemas', 'The Intersection Of Alessandro + Perris Blvds, Moreno Valley', '(909) 485-2899', 'Associate, The', '4:20 PM', 'PG-13', 'A', 'Eli Wallach'),
 ('Akarakian Theatres Moreno 4 Cinemas', 'The Intersection Of Alessandro + Perris Blvds, Moreno Valley', '(909) 485-2899', 'Associate, The', '4:20 PM', 'PG-13', 'A', 'Kenny Kerr'),
 ('Akarakian Theatres Moreno 4 Cinemas', 'The Intersection Of Alessandro + Perris Blvds, Moreno Valley', '(909) 485-2899', 'Associate, The', '4:20 PM', 'PG-13', 'A', 'Lainie Kazan'),
 ('Akarakian Theatres Moreno 4 Cinemas', 'The Intersection Of Alessandro + Perris Blvds, Moreno Valley', '(909) 485-2899', 'Associate, The', '4:20 PM', 'PG-13', 'A', 'Tim Daly'),
 ('Akarakian Theatres Moreno 4 Cinemas', 'The Intersection Of Alessandro + Perris Blvds, Moreno Valley', '(909) 485-2899', 'Associate, The', '4:20 PM', 'PG-13', 'A', 'Whoopi Goldberg'),
 ('Akarakian Theatres Moreno 4 Cinemas', 'The Intersection Of Alessandro + Perris Blvds, Moreno Valley', '(909) 485-2899', 'Associate, The', '4:20 PM', 'PG-13', 'D', 'Donald Petrie'),
 ('Akarakian Theatres Moreno 4 Cinemas', 'The Intersection Of Alessandro + Perris Blvds, Moreno Valley', '(909) 485-2899', 'Associate, The', '7:20 PM', 'PG-13', 'A', 'Austin Pendleton')]

2. **Populate the tables with data.**  

    - Use `INSERT ... SELECT ...` queries to extract data from `DATASET` for each table of your other tables. 
    - Start with the strong entities before moving on to the weak ones. Be sure to always select distinct     values. 
    - Foreign Keys get special treatment. You will have to use JOINs to pick up the PKs from the other side of the relationships. (As in Part 1, use the fact that movie names, theater names, and people names are always unique in our source data. That way you can use then for the JOIN columns. 
    - If you have to start over, then rerun your DDL queries from Part 2. 

In [26]:
%%sql
DELETE FROM THEATERS; 
INSERT INTO THEATERS (NAME, LOCATION, PHONE)
    SELECT DISTINCT TName, Location, Phone
    FROM DATASET;
SELECT * FROM THEATERS;
    
DELETE FROM MOVIES;
INSERT INTO MOVIES(TITLE,RATING)
    SELECT DISTINCT MTitle, Rating 
    FROM DATASET;
SELECT * FROM MOVIES;

DELETE FROM ARTISTS;
INSERT INTO ARTISTS (NAME)
    SELECT DISTINCT CNAME 
    FROM DATASET;
SELECT * FROM ARTISTS;

DELETE FROM CREDITS;
INSERT INTO CREDITS (CCODE, MID, ARTIST_ID)
    SELECT DISTINCT CCODE, MID, AID
    FROM DATASET 
        JOIN MOVIES ON (MOVIES.TITLE = DATASET.MTITLE)
        JOIN ARTISTS ON (ARTISTS.NAME = DATASET.CNAME);
SELECT * FROM CREDITS;

DELETE FROM SHOWS;
INSERT INTO SHOWS (ShowTime, MID, TID)
    SELECT DISTINCT ShowTime, MID, TID
    FROM DATASET
        JOIN MOVIES ON (MOVIES.TITLE = DATASET.MTITLE)
        JOIN THEATERS ON (THEATERS.NAME = DATASET.TNAME);
SELECT * FROM SHOWS;


 * sqlite:///Movies.db
9 rows affected.
9 rows affected.
Done.
23 rows affected.
23 rows affected.
Done.
153 rows affected.
153 rows affected.
Done.
162 rows affected.
162 rows affected.
Done.
0 rows affected.
131 rows affected.
Done.


[(1, '4:20 PM', 1, 1),
 (2, '7:20 PM', 1, 1),
 (3, '9:40 PM', 1, 1),
 (4, '5:10 PM', 2, 1),
 (5, '9:00 PM', 2, 1),
 (6, '9:05 PM', 2, 1),
 (7, '7:00 PM', 3, 1),
 (8, '4:40 PM', 1, 2),
 (9, '7:30 PM', 1, 2),
 (10, '2:30 PM', 4, 2),
 (11, '5:15 PM', 5, 2),
 (12, '7:45 PM', 5, 2),
 (13, '5:30 PM', 6, 2),
 (14, '8:00 PM', 6, 2),
 (15, '5:00 PM', 2, 2),
 (16, '7:40 PM', 2, 2),
 (17, '4:00 PM', 7, 2),
 (18, '6:15 PM', 7, 2),
 (19, '8:30 PM', 7, 2),
 (20, '8:20 PM', 8, 2),
 (21, '4:45 PM', 9, 2),
 (22, '5:45 PM', 9, 2),
 (23, '7:45 PM', 9, 2),
 (24, '8:45 PM', 9, 2),
 (25, '4:15 PM', 10, 2),
 (26, '4:45 PM', 10, 2),
 (27, '5:15 PM', 10, 2),
 (28, '7:00 PM', 10, 2),
 (29, '7:30 PM', 10, 2),
 (30, '8:00 PM', 10, 2),
 (31, '9:45 PM', 10, 2),
 (32, '7:30 PM', 11, 2),
 (33, '5:00 PM', 12, 2),
 (34, '6:00 PM', 12, 2),
 (35, '7:15 PM', 12, 2),
 (36, '8:15 PM', 12, 2),
 (37, '6:00 PM', 13, 2),
 (38, '5:00 PM', 1, 3),
 (39, '7:30 PM', 1, 3),
 (40, '4:45 PM', 5, 3),
 (41, '7:15 PM', 5, 3),
 (42, '5:30 PM', 6, 3),
 (43, '8:00 PM', 6, 3),
 (44, '4:00 PM', 14, 3),
 (45, '7:15 PM', 14, 3),
 (46, '5:15 PM', 15, 3),
 (47, '7:45 PM', 15, 3),
 (48, '7:00 PM', 11, 3),
 (49, '10:05 PM', 2, 4),
 (50, '5:15 PM', 2, 4),
 (51, '7:40 PM', 2, 4),
 (52, '4:40 PM', 16, 5),
 (53, '7:10 PM', 16, 5),
 (54, '9:40 PM', 16, 5),
 (55, '3:50 PM', 17, 5),
 (56, '6:50 PM', 2, 5),
 (57, '9:15 PM', 2, 5),
 (58, '6:10 PM', 3, 5),
 (59, '9:10 PM', 3, 5),
 (60, '6:20 PM', 14, 5),
 (61, '9:10 PM', 14, 5),
 (62, '4:25 PM', 18, 5),
 (63, '8:05 PM', 18, 5),
 (64, '6:20 PM', 19, 5),
 (65, '9:15 PM', 19, 5),
 (66, '4:30 PM', 15, 5),
 (67, '7:00 PM', 15, 5),
 (68, '9:30 PM', 15, 5),
 (69, '4:10 PM', 20, 5),
 (70, '6:30 PM', 20, 5),
 (71, '9:00 PM', 20, 5),
 (72, '6:40 PM', 21, 5),
 (73, '9:25 PM', 21, 5),
 (74, '6:00 PM', 22, 5),
 (75, '4:15 PM', 4, 6),
 (76, '6:40 PM', 5, 6),
 (77, '9:00 PM', 5, 6),
 (78, '4:20 PM', 6, 6),
 (79, '6:30 PM', 6, 6),
 (80, '8:40 PM', 6, 6),
 (81, '4:50 PM', 17, 6),
 (82, '7:20 PM', 2, 6),
 (83, '9:50 PM', 2, 6),
 (84, '4:30 PM', 15, 6),
 (85, '7:00 PM', 15, 6),
 (86, '9:30 PM', 15, 6),
 (87, '4:25 PM', 20, 6),
 (88, '6:50 PM', 20, 6),
 (89, '9:15 PM', 20, 6),
 (90, '5:30 PM', 11, 6),
 (91, '8:30 PM', 11, 6),
 (92, '4:40 PM', 23, 6),
 (93, '7:15 PM', 23, 6),
 (94, '9:40 PM', 23, 6),
 (95, '4:10 PM', 9, 7),
 (96, '6:45 PM', 9, 7),
 (97, '9:25 PM', 9, 7),
 (98, '4:20 PM', 10, 7),
 (99, '5:20 PM', 10, 7),
 (100, '7:00 PM', 10, 7),
 (101, '8:00 PM', 10, 7),
 (102, '9:40 PM', 10, 7),
 (103, '4:45 PM', 12, 7),
 (104, '5:50 PM', 12, 7),
 (105, '7:15 PM', 12, 7),
 (106, '8:10 PM', 12, 7),
 (107, '9:30 PM', 12, 7),
 (108, '4:20 PM', 6, 8),
 (109, '7:10 PM', 6, 8),
 (110, '9:30 PM', 6, 8),
 (111, '5:00 PM', 7, 8),
 (112, '7:00 PM', 7, 8),
 (113, '9:00 PM', 7, 8),
 (114, '4:30 PM', 15, 8),
 (115, '7:00 PM', 15, 8),
 (116, '9:45 PM', 15, 8),
 (117, '5:25 PM', 12, 8),
 (118, '7:30 PM', 12, 8),
 (119, '9:35 PM', 12, 8),
 (120, '5:15 PM', 5, 9),
 (121, '9:25 PM', 5, 9),
 (122, '7:15 PM', 2, 9),
 (123, '7:30 PM', 8, 9),
 (124, '10:20 PM', 9, 9),
 (125, '4:30 PM', 9, 9),
 (126, '7:40 PM', 9, 9),
 (127, '10:05 PM', 20, 9),
 (128, '4:40 PM', 20, 9),
 (129, '7:20 PM', 20, 9),
 (130, '4:20 PM', 11, 9),
 (131, '9:30 PM', 11, 9)]

3. **Check your work.**  
There are 9 theaters, 23 movies, 152 artists, 131 shows, and 161 credits.

In [ ]:
YOUR CODE HERE

4. **Now check the source data.**  
There are actually 13 duplicate rows in the original source data! Can you spot them? Write a SQL query that counts how many times each (MTitle, TName, Showtime, CCode, and CName) appears. Which ones appear more than once?

In [ ]:
YOUR CODE HERE

5. **Drop the `DATASET` table.**  
We don't need it anymore. 

6. **Save, commit, and push your work to GitHub.**  
Use the commit summary comment "Completed Part 3".